File: backend/src/main/java/com/smartcampost/backend/model/Payment.java
package com.smartcampost.backend.model;

import com.smartcampost.backend.model.enums.PaymentMethod;
import com.smartcampost.backend.model.enums.PaymentStatus;
import jakarta.persistence.*;
import lombok.*;

import java.time.Instant;
import java.util.UUID;

@Entity
@Table(name = "payment")
@Data
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Payment {

    @Id
    @Column(name = "payment_id", nullable = false, columnDefinition = "BINARY(16)")
    private UUID id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(
            name = "parcel_id",
            nullable = false,
            referencedColumnName = "parcel_id",
            foreignKey = @ForeignKey(name = "fk_payment_parcel")
    )
    private Parcel parcel;

    @Column(name = "amount", nullable = false)
    private Double amount;

    @Column(name = "currency", nullable = false, length = 10)
    @Builder.Default
    private String currency = "XAF";

    @Enumerated(EnumType.STRING)
    @Column(name = "method", nullable = false)
    private PaymentMethod method; // CASH, MOBILE_MONEY, CARD

    @Enumerated(EnumType.STRING)
    @Column(name = "status", nullable = false)
    private PaymentStatus status; // PENDING, SUCCESS, FAILED, CANCELLED

    @Column(
            name = "timestamp",
            nullable = false,
            updatable = false,
            columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP"
    )
    private Instant timestamp;

    @Column(name = "external_ref", length = 100)
    private String externalRef;

    // ============================================================
    // 🔥 NEW FIELD: for chargeback / reverse payment handling
    // ============================================================

    @Column(name = "reversed", nullable = false)
    @Builder.Default
    private Boolean reversed = false;

    public Boolean isReversed() {
        return reversed;
    }

    public void setReversed(Boolean reversed) {
        this.reversed = reversed;
    }

    // ============================================================
    // LIFECYCLE HOOK
    // ============================================================
    @PrePersist
    void onCreate() {
        if (id == null) {
            id = UUID.randomUUID();
        }
        if (timestamp == null) {
            timestamp = Instant.now();
        }
        if (currency == null) {
            currency = "XAF";
        }
        if (status == null) {
            status = PaymentStatus.PENDING;
        }
        if (reversed == null) {
            reversed = false;
        }
    }
}

File: smartcampost_mobile/lib/models/payment.dart
class Payment {
  final String id;
  final String? parcelId;
  final String? parcelTrackingRef;
  final double amount;
  final String? currency;
  final String? method;
  final String? status;
  final bool? reversed;
  final String? externalRef;
  final String? timestamp;

  Payment({
    required this.id,
    this.parcelId,
    this.parcelTrackingRef,
    required this.amount,
    this.currency,
    this.method,
    this.status,
    this.reversed,
    this.externalRef,
    this.timestamp,
  });

  factory Payment.fromJson(Map<String, dynamic> json) => Payment(
    id: json['id']?.toString() ?? '',
    parcelId: json['parcelId']?.toString(),
    parcelTrackingRef: json['parcelTrackingRef'] as String?,
    amount: (json['amount'] as num?)?.toDouble() ?? 0.0,
    currency: json['currency'] as String?,
    method: json['method'] as String?,
    status: json['status'] as String?,
    reversed: json['reversed'] as bool?,
    externalRef: json['externalRef'] as String?,
    timestamp: json['timestamp'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    if (parcelId != null) 'parcelId': parcelId,
    'amount': amount,
    if (currency != null) 'currency': currency,
    if (method != null) 'method': method,
    if (status != null) 'status': status,
    if (externalRef != null) 'externalRef': externalRef,
  };
}
